scalafix-nsc/src/main/scala/scalafix/nsc/NscSemanticApi.scala

```diff
@@ -11,110 +11,466 @@ import scalafix.ScalafixConfig
+
+      val refNoThis = stripThis(toShorten)
+      val names = refNoThis.collect {
+        case tn: m.Term.Name => tn
+        case tn: m.Type.Name => tn
+      }
+
+      // Mix local scope with root scope for FQN and non-FQN lookups
+      val wholeScope = mixScopes(inScope, realRootScope)
+      val symbols = lookupSymbols(names, wholeScope, toShorten.syntax)
+      val metaToSymbols = names.zip(symbols)
+
+      if (symbols.isEmpty) refNoThis
+      else {
+        val maybePathDependentType =
+          metaToSymbols.find(ms => ms._2.isAccessor)
+        val isPathDependent = maybePathDependentType.isDefined
+
+        // Get the latest value in path if it's path dependent type
+        val firstTermInPathDependent = maybePathDependentType.flatMap(_ =>
+          metaToSymbols.dropWhile(ms => !ms._2.isValue).headOption)
+
+        val (lastName, lastSymbol) =
+          firstTermInPathDependent.getOrElse(metaToSymbols.last)
+        val (onlyPaths, shortenedNames) =
+          metaToSymbols.span(_._1 != lastName)
+
+        // Build map of meta names to reflect names
+        val renames: Map[m.Name, g.Name] = metaToSymbols.map {
+          case (metaName, symbol) =>
+            val mappedSym = st.renames.getOrElse(symbol, symbol)
+            metaName -> mappedSym.name
+        }.toMap
+
+        // Get shortened and renamed type name
+        val onlyShortenedNames = shortenedNames.map(_._1)
+        val typeName = toTypeName(onlyShortenedNames.last)
+        val renamedTypeName = renameType(typeName, renames)
+
+        /* Strategy:
+         *   - Has the local scope lookup succeded and is not a PDT?
+         *     Return type name since it's already in the scope.
+         *   - Otherwise, compute the shortened prefix to use. */
+        val localSym = inScope.lookup(lastSymbol.name)
+        if (!isPathDependent && localSym.exists) renamedTypeName
+        else {
+          // Compute the shortest prefix that we append to type name
+          val shortenedRefs = {
+            if (isPathDependent) onlyShortenedNames.init
+            else {
+              // Get last symbol that was found in scope
+              val lastAccessibleRef =
+                onlyPaths.reduceLeft[(m.Name, g.Symbol)] {
+                  case (t1 @ (_, s1), t2 @ (_, s2)) =>
+                    if (s1.exists && s2.exists) t2 else t1
+                }
+              val lastName = lastAccessibleRef._1
+              renameType(lastName, renames) :: onlyShortenedNames.init
+            }
+          }
+          val termRefs = shortenedRefs.map(toTermName)
+          val termSelects = termRefs.reduceLeft[m.Term.Ref] {
+            case (qual: m.Term, path: m.Term.Name) =>
+              m.Term.Select(qual, renameType(path, renames))
+          }
+          m.Type.Select(termSelects, renamedTypeName)
+        }
+      }
+    }
+
     new SemanticApi {
+      override def shortenType(tpe: m.Type, owner: m.Tree): m.Type = {
```

Comment:
This signature is tricky because the current impl only works for defn.{val,var,def}, so  the `owner` should really be a `Defn`, if anything. With a position, it wouldn't be clear which position you should pass in e.g. the defn's name or the defn itself.

IIRC, the `owner` needs to map to some tree with an ownerchain.